src/Microsoft.ML.Transforms/Text/NgramTransform.cs

```diff
@@ -178,350 +244,115 @@ public void Save(ModelSaveContext ctx)
-                        });
-                }
-            }
-        }
-
-        private static string TestType(ColumnType type)
+        private static (string input, string output)[] GetColumnPairs(ColumnInfo[] columns)
         {
-            const string reason = "Expected vector of Key type, and Key is convertable to U4";
-            Contracts.AssertValue(type);
-            if (!type.IsVector)
-                return reason;
-            if (!type.ItemType.IsKey)
-                return reason;
-            // Can only accept key types that can be converted to U4.
-            if (type.ItemType.KeyCount == 0 && type.ItemType.RawKind > DataKind.U4)
-                return reason;
-            return null;
+            Contracts.CheckValue(columns, nameof(columns));
+            return columns.Select(x => (x.Input, x.Output)).ToArray();
         }
 
-        private void InitColumnTypeAndMetadata(out VectorType[] types, out VectorType[] slotNamesTypes)
+        protected override void CheckInputColumn(ISchema inputSchema, int col, int srcCol)
         {
-            types = new VectorType[Infos.Length];
-            slotNamesTypes = new VectorType[Infos.Length];
-
-            var md = Metadata;
-            for (int iinfo = 0; iinfo < _exes.Length; iinfo++)
-            {
-                var vectorSize = _ngramMaps[iinfo].Count;
-                if (vectorSize == 0)
-                {
-                    // If we the ngram map is empty, we should output the fixed-size vector of size 0. This is not a
-                    // valid VectorType (it would be considered variable), so we make it size 1 in this case.
-                    vectorSize = 1;
-                }
-                types[iinfo] = new VectorType(NumberType.Float, vectorSize);
-                var info = Infos[iinfo];
-                if (!Source.Schema.HasKeyNames(info.Source, info.TypeSrc.ItemType.KeyCount))
-                    continue;
-
-                using (var bldr = md.BuildMetadata(iinfo))
-                {
-                    if (_ngramMaps[iinfo].Count > 0)
-                    {
-                        slotNamesTypes[iinfo] = new VectorType(TextType.Instance, _ngramMaps[iinfo].Count);
-                        bldr.AddGetter<VBuffer<ReadOnlyMemory<char>>>(MetadataUtils.Kinds.SlotNames,
-                            slotNamesTypes[iinfo], GetSlotNames);
-                    }
-                }
-            }
-            md.Seal();
+            var type = inputSchema.GetColumnType(srcCol);
+            if (!NgramEstimator.IsColumnTypeValid(type))
+                throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", ColumnPairs[col].input, NgramEstimator.ExpectedColumnType, type.ToString());
         }
 
-        private void GetSlotNames(int iinfo, ref VBuffer<ReadOnlyMemory<char>> dst)
+        public NgramTransformer(IHostEnvironment env, IDataView input, ColumnInfo[] columns)
```

Comment:
>public [](start = 8, length = 6)

training ctor should not be public

Reply:
I would make an internal static training method

---
In reply to: [233694151](https://github.com/dotnet/machinelearning/pull/1592#discussion_r233694151) [](ancestors = 233694151)